File: 017.zig
// NOTE: there's probably another way
// rather than comptime exploiting. :P
const std = @import("std");
const stdout = std.io.getStdOut().writer();

fn unit(comptime n: u64) []const u8 {
    switch (n) {
        0 => return "zero",
        1 => return "one",
        2 => return "two",
        3 => return "three",
        4 => return "four",
        5 => return "five",
        6 => return "six",
        7 => return "seven",
        8 => return "eight",
        9 => return "nine",
        else => return "unknown",
    }
}

fn uniqueTwoDigit(comptime n: u64) []const u8 {
    switch (n) {
        10 => return "ten",
        11 => return "eleven",
        12 => return "twelve",
        13 => return "thirteen",
        14 => return "fourteen",
        15 => return "fifteen",
        16 => return "sixteen",
        17 => return "seventeen",
        18 => return "eighteen",
        19 => return "nineteen",
        else => return "unknown",
    }
}

fn twoDigit(comptime n: u64) []const u8 {
    const prefixes = comptime [8][]const u8{
        "twenty","thirty", "forty",
        "fifty", "sixty", "seventy",
        "eighty", "ninety",
    };

    const first = comptime n/10;
    const second = comptime n%10;

    if (first == 0) {
        return unit(second);
    } else if (first == 1 and (second >= 0 and second <= 9)) {
        return uniqueTwoDigit(n);
    } else if (second == 0) {
        return prefixes[first - 2];
    } else {
        const single = comptime unit(second);
        return prefixes[first - 2] ++ single;
    }
}

fn threeDigit(comptime n: u64) []const u8 {
    const first = comptime n/100;
    const last_two = comptime n%100;

    comptime var num = unit(first) ++ "hundred";
    if (last_two == 0) {
        return num;
    } else {
        return num ++ "and" ++ comptime twoDigit(last_two);
    }
}

fn nToWord(comptime n: u64) []const u8 {
    switch (n) {
        0...9 => return unit(n),
        10...19 => return uniqueTwoDigit(n),
        20...99 => return twoDigit(n),
        100...999 => return threeDigit(n),
        1000 => return "onethousand",
        else => return "unknown",
    }
}

fn solve(comptime n: u64) u64 {
    comptime var i: u64 = 1;
    var total: u64 = 0;
    inline while (i <= n) : (i += 1) { // For some reason you cannot exceed 1000 loops. Lol!
        total += nToWord(i).len;
    }

    return total;
}

pub fn main() !void {
    try stdout.print("{}\n", .{solve(1000)});
}

File: 017.c
// 21124
#include <stdio.h>

#define AND 3
#define ONETHOUSAND 11

int one_digit(int);
int two_digit(int);
int three_digit(int);
int compute_sum(void);

int main()
{
    printf("%d\n", compute_sum());
    return 0;
}

int one_digit(int n)
{
    int lens[9] = {
        3, 3, 5, 4, 4, 3, 5, 5, 4
    };
    return lens[n - 1];
}

int two_digit(int n)
{
    /* multiples of ten */
    int tens[9] = {
        3, 6, 6, 5, 5, 5, 7, 6, 6
    };

    /* eleven to nineteen */
    int unique[9] = {
        6, 6, 8, 8, 7, 7, 9, 8, 8
    };
    if (n % 10 == 0)
        return tens[n/10 - 1];
    else if (n >= 11 && n <= 19)
        return unique[n - 11];
    else {
        int d1 = n/10;
        int d2 = n%10;
        /* consider 01, 02... */
        if (d1 == 0)
            return one_digit(d2);

        return tens[d1 - 1] + one_digit(d2);
    }
}

int three_digit(int n)
{
    /* multiples of 100 */
    int hundreds[9] = {
        10, 10, 12, 11, 11, 10, 12, 12, 11
    };
    if (n % 100 == 0)
        return hundreds[n/100 - 1];
    else {
        int mul = n/100;
        int n2 = n%100;
        return hundreds[mul - 1] + AND + two_digit(n2);
    }
}

int compute_sum(void)
{
    int sum, i;

    for (sum = 0, i = 1; i < 1000; ++i) {
        if (i >= 1 && i <= 9)
            sum += one_digit(i);
        else if (i >= 10 && i <= 99)
            sum += two_digit(i);
        else if (i >= 100 && i <= 999)
            sum += three_digit(i);
    }
    sum += ONETHOUSAND;
    return sum;
}
